vispy/visuals/volume.py

```diff
@@ -275,6 +286,65 @@
 """  # noqa
 
 
+RAYCASTING_SETUP_VOLUME = """
+    // Compute the distance to the front surface or near clipping plane
+    float distance = dot(nearpos-v_position, view_ray);
+    distance = max(distance, min((-0.5 - v_position.x) / view_ray.x,
+                            (u_shape.x - 0.5 - v_position.x) / view_ray.x));
+    distance = max(distance, min((-0.5 - v_position.y) / view_ray.y,
+                            (u_shape.y - 0.5 - v_position.y) / view_ray.y));
+    distance = max(distance, min((-0.5 - v_position.z) / view_ray.z,
+                            (u_shape.z - 0.5 - v_position.z) / view_ray.z));
+
+    // Now we have the starting position on the front surface
+    vec3 front = v_position + view_ray * distance;
+
+    // Decide how many steps to take
+    int nsteps = int(-distance / u_relative_step_size + 0.5);
+    float f_nsteps = float(nsteps);
+    if( nsteps < 1 )
+        discard;
+
+    // Get starting location and step vector in texture coordinates
+    vec3 step = ((v_position - front) / u_shape) / f_nsteps;
+    vec3 start_loc = front / u_shape;
+"""
+
+RAYCASTING_SETUP_PLANE = """
+    // find intersection of view ray with plane in data coordinates
+    vec3 intersection = intersectLinePlane(v_position.xyz, view_ray, 
+                                           u_plane_position, u_plane_normal);
+    // and texture coordinates
+    vec3 intersection_tex = intersection / u_shape;
+    
+    // discard if intersection not in texture
+    if (intersection_tex.x > 1.0 )
+        discard;
+    if (intersection_tex.y > 1.0 )
+        discard;
+    if (intersection_tex.z > 1.0 )
+        discard;
+    if (intersection_tex.x < 0.0 )
+        discard;
+    if (intersection_tex.y < 0.0 )
+        discard;
+    if (intersection_tex.z < 0.0 )
+        discard;
```

Comment:
@almarklein You've been a real stickler on conditionals in shader code, any idea if there is a preferred way of handling the above? Could you do `intersection_tex.xyz < 0`? Or I suppose that is an AND when we need an OR.

Reply:
would love to hear if there's a better way of doing this - for context, I implemented as a series of `if` statements to allow for the earliest possible exit

Reply:
My gut says that many if-statements look scary :)  I would feel more confortable if we can convert this to a single `if` and a bunch of OR's and AND's. But also it's 2021 and I'm not sure how much this matters on modern hardware. And whether replacing many with (still) one helps.

@rougier what are your thoughts on this? Similar questions have popped up before. Is there a resource that we could use as a guideline when dealing with conditionals in shaders (in this day and age)?

Reply:
My laziness says this won't matter overall (it won't be noticeable to the majority of users). My developer gut says these should at least be combined for cleanliness so the `.x` comparisons happen in one if statement, the `.y` in one if, and the `.z` in the third and last if statement. @alisterburt would you mind updating this to do that.

Reply:
implemented an alternative suggestion from @kevinyamauchi to reduce the branching

```
    // discard if intersection not in texture
    
    float out_of_bounds = 0;

    out_of_bounds += float(intersection_tex.x > 1);
    out_of_bounds += float(intersection_tex.x < 0);
    out_of_bounds += float(intersection_tex.y > 1);
    out_of_bounds += float(intersection_tex.y < 0);
    out_of_bounds += float(intersection_tex.z > 1);
    out_of_bounds += float(intersection_tex.z < 0);
    
    if (out_of_bounds > 0) {
        discard;
    }
```

hope that's okay!